### ISSUE TITLE
Problems with OSWEB together with JATOS on a server

### ISSUE THREAD

**Issue by kristian-lange** [shyras/osweb#120]
I'm not sure whether this is actually an OSWEB issue, but anyway even if not, maybe you have an idea what the underlying reason is. 

The experiment was created with OSWEB 1.2.4 (actually not by me). It was imported into an local JATOS and there it was running fine.
Then, imported into an server running in the cloud (Digital Ocean), it stopped working: Mostly it displays only a white screen and in the console are errors:

    NotSupportedError: Operation is not supported jatos.js:80
    TypeError: alertify.errorAlert is not a function[Learn More] jatos.js:80:5
    TypeError: this._container is null[Learn More] osweb.1.2.4.bundle.js:1:168886

But the weird thing is that sometimes with an reload this problem vanishes and the experiment works. I couldn't find the actual cause for this but I found a (ugly) way to circumvent it: Loading the OSWEB libraries delayed by 2 s helps. Just move the loading from the <head> to a `setTimeout` seems to solve the immediate problem:

```javascript
    <script>
      setTimeout(function(){
        $.getScript( "js/vendors~osweb.1.2.4.bundle.js" )
        .then(function(){
          return $.getScript( "js/osweb.1.2.4.bundle.js" )
        })
        .then(function(){
          return $.getScript( "js/jatos2.js" )
        })
      }, 2000);
    </script>
```

Now, does anyone has an idea what is going on here?

**Comment by smathot** [comment]
It looks like the last error (`this._container is null`) is the real cause. But it's hidden, because the other two errors arise when OSWeb tries to notify the user through a popup with `alertify.errorAlert()`. In very general terms though, the fact that the issue goes away when running locally and when adding a delay suggests that it has to do with loading times.

A few questions:

- Is the experiment file large? (I.e. how many MB)
- Does the issue also occur if you run a very basic experiment, say the default template?

**Comment by kristian-lange** [comment]
_Since I'm not the actual experimenter I'm forwarding the answer here:_

The experimental file is 5.217 kb. I reduced some of the stimuli files, resulting in a version that is only 2.811 kb and the same problems showed up (again only on the server.. = white screen, the experiment does not start). When reducing the whole experiment even more (by only showing an extremely basic version = sketchpads with text which again works fine offline), the problem also remains..

I also tried out your fixed version again. It seems to run fine with some browsers but less so with others (e.g. after a complete run of the whole study in google chrome, the following would appear on the screen: 
"A problem occurred:
Errors occurred during the experiment").

**Comment by smathot** [comment]
I'll see if I can replicate this. @dschreij Do you have any idea what the issue might be?

**Comment by RichardTindle** [comment]
I am having a similar issue. I am able to run the experiment on JATOS locally but once I generate a link and try to start on another device it does not work.
On the other device I get 

"The web page at...could not be loaded becuase: net::ERR_CONNECTION_REFUSED

When I run the loader.bat it displays this:
No Instance(s) Available.
Starting JATOS ... please wait
JATOS uses local JRE
To use JATOS type 127.0.0.1:9000 in your browser's address bar
Picked up JAVA_TOOL_OPTIONS: -Djava.vendor="Sun Microsystems Inc."
NettyServer.main is deprecated. Please start your Play server with the play.core.server.ProdServerStart$.main.

**Comment by kristian-lange** [comment]
Hi Richard,

now that I read your message on GitHub it looks more like an JATOS issue
and not like one from OSWEB. Sorry about that.
This error msg hints at some network problems:
"The web page at...could not be loaded becuase: net::ERR_CONNECTION_REFUSED

Can you give me more details about your setup: Where is your JATOS server
running (local server, cloud)? What OS has your server? What is your IP and
port? Which version is your JATOS?
Also, can you send me JATOS' logs? They are in logs/application.logs. If it
is too long, I'm mostly interested in the startup phase.

Best,
Kristian

**Comment by kristian-lange** [comment]
@RichardTindle Your issue looks more like a JATOS issue. Let's keep issues separated. I wrote you an email.

**Comment by dschreij** [comment]
@RichardTindle. I think this is because your JATOS server only listens from connections from the local network, judging from 127.0.0.1:9000. If it also listens for connections from outside, this usually says 0.0.0.0:9000, or something similar.

**Comment by dschreij** [comment]
> I'll see if I can replicate this. @dschreij Do you have any idea what the issue might be?

Not at first sight. Errors occurred during the experiment is too generic to tell us anything. My hunch is that there is some race condition going on, where some libraries are loaded too fast, while they depend on libraries that load a bit later. This is weird, because we have placed all Osweb code in the jatos.onLoad() callback, which should prevent this from happening.
Moreover, our installation of Jatos also runs on a VPS (from TransIP) and we have not experienced any problems like these, so I have no idea at the moment about how to try to replicate this issue.

**Comment by smathot** [comment]
I haven't seen the issue before either, nor have I been able to replicate it. Let's treat it as a rare edge case for now (related to either the experiment or the server, that I'm not sure about), but keep an eye out for it in the future.

**Comment by kristian-lange** [comment]
May I weigh in here: It's running on a standard Digital Ocean instance (they call it Droplet). Nothing unusual there. Sarah, who created the experiment, gave me an user account for this JATOS and I checked it myself. It reliably didn't work.

**Comment by dschreij** [comment]
Alright, we could spawn a droplet ourselves and try it out, if need be. I know DigitalOcean allows you to dot this and pay per minute (for 1 ct per hour or so). I currently do not have much time at hand, so I can't make any promises about when I'm thoroughly able to look at this. Might be in two weeks earliest.
Could you provide us with some more info, to create the similar environment:

- What OS does the droplet host?
- What version of Jatos is she using?
- Is she using Docker to run everything or is Jatos installed locally?
- Is she using Jatos with Java packaged, or a system install of Java?
- Is she using MySQL or the companion database dat Jatos uses?
- Can we get access to her experiment?
- What are the exact steps that she takes before the problem occurs (is she using the exporter extensions for Jatos that is included in OpenSesame?)

**Comment by SarahVerlage** [comment]
Hi all, 
Yes, unfortunately it really did not work reliably on Digital Ocean. Kristian was kind enough to provide me with a fix that seems to work ok most of the time, but it would be really great if we found out what the actual problem is.. 
Would it be an option to add one of you (smathot or dschreij) as a user to my study on JATOS?
Best,
Sarah

**Comment by smathot** [comment]
@SarahVerlage Thanks for chipping in.

> Would it be an option to add one of you (smathot or dschreij) as a user to my study on JATOS?

Yes, it would! Could you create an account for both us?

**Comment by smathot** [comment]
I think I traced it down. See smathot/opensesame-extension-osweb#13. I'm not sure whether this is also a bug in JATOS, or whether this is intended behavior, but the bug appears to be due (if my analysis is correct) to the fact that `jatos.onLoad()` does not wait for the page to be loaded.

@dschreij Could you verify the pull request? Then I'll push a bugfix release of the extension (1.2.5.2).

**Comment by dschreij** [comment]
Great work! I tried logging in, but wasn't allowed with the account. I also tried using my e-mail address (as Jatos asks for) as a username instead of Daniel, but that didn't work either. I'm glad you already hunted this one down. I'll have a look at the PR now.

**Comment by kristian-lange** [comment]
@smathot It's actually intended behaviour. jatos.onLoad does wait for jatos.js to be loaded and ready to be used but not necessary for the page to be fully loaded. So far this has never been an issue. But maybe it's a good idea to do so. Let me think about it.

**Comment by smathot** [comment]
The issue should be resolved in 1.2.5.2 of the OSWeb extension (so not OSWeb itself, but rather its integration with OpenSesame and JATOS). Executing the following command in the OpenSesame debug window upgrades the extension (you possibly have to run OpenSesame as administrator on Windows). Then you can re-export the experiment and the issue should be resolved.

~~~
!pip install opensesame-extension-osweb --upgrade
~~~

Please let us know whether this indeed fixes the issue!

**Comment by kristian-lange** [comment]
Just heard from Sarah, that it's working with her now :+1:
